src/tflite_cpu.cpp

```diff
@@ -0,0 +1,756 @@
+//
+// Configuration requires the following parameters:
+//   -- `model_path`: An absolute filesystem path to a TensorFlow Lite model file.
+//
+// The following optional parameters are honored:
+//   -- `num_threads`: Sets the number of threads to be used, where applicable.
+//
+//   -- `label_path`:  An absolute filesystem path to a .txt file of the model's category labels.
+//
+// Any additional configuration fields are ignored.
+class MLModelServiceTFLite : public vsdk::MLModelService,
+                             public vsdk::Stoppable,
+                             public vsdk::Reconfigurable {
+    class write_to_tflite_tensor_visitor_;
+
+   public:
+    explicit MLModelServiceTFLite(vsdk::Dependencies dependencies,
+                                  vsdk::ResourceConfig configuration)
+        : MLModelService(configuration.name()),
+          state_(reconfigure_(std::move(dependencies), std::move(configuration))) {}
+
+    ~MLModelServiceTFLite() final {
+        // All invocations arrive via gRPC, so we know we are idle
+        // here. It should be safe to tear down all state
+        // automatically without needing to wait for anything more to
+        // drain.
+    }
+
+    void stop(const vsdk::ProtoStruct& extra) noexcept final {
+        return stop();
+    }
+
+    /// @brief Stops the MLModelServiceTFLite from running.
+    void stop() noexcept {
+        using std::swap;
+        try {
+            const std::lock_guard<std::mutex> lock(state_lock_);
+            if (!stopped_) {
+                stopped_ = true;
+                std::shared_ptr<state> state;
+                swap(state_, state);
+                state_ready_.notify_all();
+            }
+        } catch (...) {
+        }
+    }
+
+    void reconfigure(const vsdk::Dependencies& dependencies,
+                     const vsdk::ResourceConfig& configuration) final try {
+        // Care needs to be taken during reconfiguration. The
+        // framework does not offer protection against invocation
+        // during reconfiguration. Keep all state in a shared_ptr
+        // managed block, and allow client invocations to act against
+        // current state while a new configuration is built, then swap
+        // in the new state. State which is in use by existing
+        // invocations will remain valid until the clients drain. If
+        // reconfiguration fails, the component will `stop`.
+
+        // Swap out the state_ member with nullptr. Existing
+        // invocations will continue to operate against the state they
+        // hold, and new invocations will block on the state becoming
+        // populated.
+        using std::swap;
+        std::shared_ptr<state> state;
+        {
+            // Wait until we have a state in play, then take
+            // ownership, so that we don't race with other
+            // reconfigurations and so other invocations wait on a new
+            // state.
```

Comment:
I guess this works, but it's so much more complicated than it needs to be. To avoid race conditions with other reconfigurations, just lock the mutex here at the top of the function and keep it locked until we return at the bottom! No need for condition variables at all:

```c++
void reconfigure(const vsdk::Dependencies& dependencies,
                     const vsdk::ResourceConfig& configuration) final try {
        using std::swap;
        std::unique_lock<std::mutex> lock(state_lock_);
        check_stopped_inlock_();
        std::shared_ptr<state> state = reconfigure_(dependencies, configuration);
        swap(state_, state);
} catch (...) {
    ...
}
```